### PR TITLE
Add expansion of context.(pipeline|pipelinerun|task|taskRun).name and add tests

### PR DESCRIPTION
**Pull request by R2wenD2**
# Changes

Add expansion of context.(pipeline|pipelinerun|task|taskRun).name and add tests and documentation. This addresses #1522

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [ X] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```

$context.(task|taskRun|pipeline|pipelineRun).name can be used to reference the appropriate name when in pipelineRuns and taskRuns


```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 1]
```diff
@@ -53,6 +53,16 @@ func ApplyParameters(p *v1beta1.PipelineSpec, pr *v1beta1.PipelineRun) *v1beta1.
 	return ApplyReplacements(p, stringReplacements, arrayReplacements)
 }
 
+// ApplyContexts applies the substitution from $(context.(pipelineRun|pipeline).*) with the specified values.
+// Currently supports only name substitution. Uses "" as a default if name is not specified.
+func ApplyContexts(spec *v1beta1.PipelineSpec, pipelineName string, pr *v1beta1.PipelineRun) *v1beta1.PipelineSpec {
+	stringReplacements := map[string]string{}
```
Comment (ImJasonH) [inline pkg/reconciler/pipelinerun/resources/apply.go:59]: style nit: I'd just declare these in the map initialization directly:

```
stringReplacements := map[string]string{
  "context.pipelineRun.name": pr.Name,
  "context.pipeline.name": pipelineName,
}
```

(This also lets you pass the map directly to `ApplyReplacements`, and make this method a one-liner, if you want)

File: pkg/reconciler/taskrun/resources/apply.go [hunk 2]
```diff
@@ -96,6 +96,16 @@ func ApplyResources(spec *v1beta1.TaskSpec, resolvedResources map[string]v1beta1
 	return ApplyReplacements(spec, replacements, map[string][]string{})
 }
 
+// ApplyContexts applies the substitution from $(context.(taskRun|task).*) with the specified values.
+// Currently supports only name substitution. Uses "" as a default if name is not specified.
+func ApplyContexts(spec *v1beta1.TaskSpec, rtr *ResolvedTaskResources, tr *v1beta1.TaskRun) *v1beta1.TaskSpec {
+	stringReplacements := map[string]string{}
```
Comment (ImJasonH) [inline pkg/reconciler/taskrun/resources/apply.go:102]: ditto

File: pkg/reconciler/taskrun/resources/apply.go [hunk 3]
```diff
@@ -96,6 +96,16 @@ func ApplyResources(spec *v1beta1.TaskSpec, resolvedResources map[string]v1beta1
...
+	stringReplacements["context.taskRun.name"] = tr.Name
+	stringReplacements["context.task.name"] = rtr.TaskName
```
Comment (ImJasonH) [inline pkg/reconciler/taskrun/resources/apply.go:104]: Just to double-check, if I'm defining a TaskRun's TaskSpec inline, there won't be a Task name, and this context value will be `""`, right?

Comment (R2wenD2) [reply pkg/reconciler/taskrun/resources/apply.go:104]: yes, and I added a testcase to confirm.

**Pull request merged**